### PR TITLE
Increase Cypress tests to check reset code and loop through all lesson types

### PR DESCRIPTION
**Pull request by terichadbourne**
@zebateira Can you please help me figure out why Cypress is unable to run after this last commit? 

The first commit refactored so that lesson and tutorials types were reported directly and only the TypeIcon component clumps `code` together with `fileUpload`. 

The second commit added testing for the reset code functionality, which was passing. 

In the third commit I was trying to refactor so that we loop through every lesson in every tutorial and detect its type to decide what tests to run on it, with some cheats built in to advance past `file-upload` and `multiple-choice` lessons for now. (This is as opposed to feeding it a list of standard coding tutorials and only testing those, ignoring the others.) 

In its current state the site appears to run just fine for me with no errors in Terminal, but Cypress bails out before running any tests. The only changes in that 3rd commit were to the tests, not the main site, so I'm not sure why it's saying there's something wrong outside the tests. 

If I do just `npm run serve` the only console errors I see are many, many of this one while in some lesson files, but I'm not sure if it was already there when the 2nd commit passed the Cypress tests.

image

Cypress screenshot from `npm run test`: 

image

Here's what I see in Terminal after running `npm run test`: 

```sh
~/code/pl/protoschool/proto.school (test-reset-code) ! $ npm run test

> protoschool@1.0.0 test /Users/terichadbourne/code/pl/protoschool/proto.school
> start-server-and-test serve http://localhost:3000 cy:run

starting server using command "npm run serve"
and when url "http://localhost:3000" is responding
running tests using command "npm run cy:run"

> protoschool@1.0.0 serve /Users/terichadbourne/code/pl/protoschool/proto.school
> vue-cli-service serve

 INFO  Starting development server...
98% after emitting CopyPlugin

 DONE  Compiled successfully in 8762ms                                               16:27:14


  App running at:
  - Local:   http://localhost:3000 
  - Network: http://192.168.12.247:3000

  Note that the development build is not optimized.
  To create a production build, run npm run build.


> protoschool@1.0.0 cy:run /Users/terichadbourne/code/pl/protoschool/proto.school
> cypress run --browser chrome


=============================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:    3.8.3                                                                              │
  │ Browser:    Chrome 80                                                                          │
  │ Specs:      1 found (tutorials.spec.js)                                                        │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  tutorials.spec.js                                                               (1 of 1)


  1) An uncaught error was detected outside of a test

  0 passing (786ms)
  1 failing

  1)  An uncaught error was detected outside of a test:
     Uncaught TypeError: Cannot convert undefined or null to object

This error originated from your test code, not from Cypress.

When Cypress detects uncaught errors originating from your test code it will automatically fail the current test.

Cypress could not associate this error to any specific test.

We dynamically generated a new test to display this failure.
      at Function.keys (<anonymous>)
      at Suite.<anonymous> (http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:8:364611)
      at Module.<anonymous> (http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:8:364535)
      at n (http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:1:110)
      at Object.<anonymous> (http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:8:108314)
      at n (http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:1:110)
      at http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:1:902
      at http://localhost:3000/__cypress/tests?p=cypress/integration/tutorials.spec.js-956:1:912




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        1                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        true                                                                             │
  │ Duration:     0 seconds                                                                        │
  │ Spec Ran:     tutorials.spec.js                                                                │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /Users/terichadbourne/code/pl/protoschool/proto.school/cypress/screenshots/tutor    (2400x2838)
     ials.spec.js/An uncaught error was detected outside of a test (failed).png                     


  (Video)

  -  Started processing:  Compressing to 32 CRF                                                     
  -  Finished processing: /Users/terichadbourne/code/pl/protoschool/proto.school/cypr    (3 seconds)
                          ess/videos/tutorials.spec.js.mp4                                          


=============================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  tutorials.spec.js                        783ms        1        -        1        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 1 failed (100%)                     783ms        1        -        1        -        -  

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! protoschool@1.0.0 cy:run: `cypress run --browser chrome`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the protoschool@1.0.0 cy:run script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/terichadbourne/.npm/_logs/2020-02-14T21_27_34_380Z-debug.log
TypeError: util.getSystemErrorName is not a function
    at getCode (/Users/terichadbourne/code/pl/protoschool/proto.school/node_modules/start-server-and-test/node_modules/execa/lib/error.js:11:16)
    at makeError (/Users/terichadbourne/code/pl/protoschool/proto.school/node_modules/start-server-and-test/node_modules/execa/lib/error.js:50:35)
    at handlePromise (/Users/terichadbourne/code/pl/protoschool/proto.school/node_modules/start-server-and-test/node_modules/execa/index.js:112:26)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! protoschool@1.0.0 test: `start-server-and-test serve http://localhost:3000 cy:run`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the protoschool@1.0.0 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/terichadbourne/.npm/_logs/2020-02-14T21_27_34_483Z-debug.log
```

These are the contents of the log file it told me to look at: 

image

### CODE REVIEW

**Comment by terichadbourne** [comment]
Thanks for the debugging help @zebateira! I was able to get this working again and do some refactoring, so this PR now adds: 
- testing of the reset code feature on all code challenge and file upload lessons
- testing of text lessons (working next button)
- looping through all lessons of every tutorial (previously only examined a set list of tutorials that had only standard code challenges

Steps taken and means of advancing to next lesson depend on the lesson type being viewed: 
- standard code challenge 
   - test reset code
   - copy solution code and submit it to pass challenge
   - click next
- text lesson 
   - click next
- file upload 
   - test reset code
   - copy solution code 
   - **_unable to submit because button is disabled pending file upload - we need to build a system for fake file upload in a future PR_**
   - **cheat to advance** in Cypress by visiting next lesson's URL
- multiple choice
   - **_unable to pass lesson - we need to build this Cypress functionality in a future PR_**
   - **cheat to advance** in Cypress by visiting next lesson's URL

Here's a sample of the loop through lessons for a tutorial with mixed lesson types: 

image

I believe this is ready to go! If you're comfortable with this as is, we can merge and I'll start a new PR to add multiple choice testing.

**Comment by terichadbourne** [comment]
@zebateira This is ready for review. Latest edits/new functionality: 
- Instead of having the separate test that clicked through every link on the landing page, we now click from landing page to first lesson, click nav back to landing page, then click our way to advance through all lessons, assuming if one lesson link from the landing page works, that functionality is working. 
- Reset code functionality is now only tested twice: once on a sample code challenge and once on a sample file upload (previously tested on every single lesson of either type
- Increased default timeout in Cypress from 4 to 8 seconds
- Added test to ensure tutorials page lists every tutorial that exists in `tutorials.json` in the order specified by the `all` array in `courses.json`
- Added test to ensure that the toggle to hide coding tutorials does its job (which is good because I discovered that I had broken that feature earlier in this PR 😂)
- Added test to ensure the homepage shows in the correct order all tutorials specified in the `featured` array in `courses.json`

File: cypress/integration/tutorials.spec.js [hunk 1]
```diff
@@ -1,79 +1,191 @@
...
+  code: {
+    tutorialId: '0002',
+    lessonNr: '02'
+  },
+  fileUpload: {
+    tutorialId: '0004',
+    lessonNr: '05'
+  },
+  // TODO: Uncomment when we have testing for multiple choice built and Anatomy of a CID is published
+  // multipleChoice: {
```
Comment (zebateira) [inline cypress/integration/tutorials.spec.js]: Ref: #359

File: cypress/integration/tutorials.spec.js [hunk 2]
```diff
@@ -1,79 +1,191 @@
...
+    }
+  })
+  it(`toggle hides coding tutorials`, function () {
+    const codelessTutorials = courses.all.filter(tutorialId => (getTutorialType(tutorialId) !== 'code') && (getTutorialType(tutorialId) !== 'file-upload'))
+    cy.log('codelessTutorials', codelessTutorials)
+    cy.get('[data-cy=toggle-coding-tutorials]').click()
+    cy.get('[data-cy=tutorial-title]').should('have.length', codelessTutorials.length) // displaying # of tutorials in tutorials.json
+    // for (let i = 0; i < codelessTutorials.length; i++) {
+    //   cy.get('[data-cy=tutorial-title]').eq(i).should('contain', tutorials[codelessTutorials[i]].title)
+    // }
```
[the remark this answers is not in the document]
Comment (terichadbourne) [reply cypress/integration/tutorials.spec.js]: Oops! Nope, meant to be used. Will uncomment.

File: cypress/integration/tutorials.spec.js [hunk 3]
```diff
@@ -1,79 +1,191 @@
...
+    cy.url().should('include', `#/${tutorialName}/01`)
+    cy.get(`[data-cy=tutorial-landing-link]`).click() // test nav link back to tutorial landing page
+    cy.contains('h2', tutorialTitle)
+    cy.get(`[href="#/${tutorialName}/01"]`).click()
+    cy.contains('h1', standardLessons[0].title)
   })
-  // loop through standard lessons and attempt to pass challenges
+
+  // ALL TUTORIAL TYPES - loop through all lessons
   for (let i = 1; i <= lessonCount; i++) {
```
[the remark this answers is not in the document]
Comment (terichadbourne) [reply cypress/integration/tutorials.spec.js]: fixed!

**Comment by terichadbourne** [comment]
@zebateira I've addressed your feedback and have also pulled out the code that checks the view solution toggle as I did earlier with the reset code toggle. Also did some general refactoring and added a better check for success on the code challenges. Let me know what you think!

**Pull request merged**